sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py

```diff
@@ -172,3 +174,152 @@ def test_binning_train_validation_are_separated():
                   int((1 - validation_fraction) * n_samples))
     assert np.all(mapper_training_data.actual_n_bins_ !=
                   mapper_whole_data.actual_n_bins_)
+
+
+@pytest.mark.parametrize('GradientBoosting, X, y', [
+    (HistGradientBoostingClassifier, X_classification, y_classification),
+    (HistGradientBoostingRegressor, X_regression, y_regression)
+])
+def test_max_iter_with_warm_start_validation(GradientBoosting, X, y):
+    # Check that a ValueError is raised when the maximum number of iterations
+    # is smaller than the number of iterations from the previous fit when warm
+    # start is True.
+
+    estimator = GradientBoosting(max_iter=50, warm_start=True)
+    estimator.fit(X, y)
+    estimator.set_params(max_iter=25)
+    err_msg = ('max_iter=25 must be larger than or equal to n_iter_=50 '
+               'when warm_start==True')
+    with pytest.raises(ValueError, match=err_msg):
+        estimator.fit(X, y)
+
+
+@pytest.mark.parametrize('GradientBoosting, X, y', [
+    (HistGradientBoostingClassifier, X_classification, y_classification),
+    (HistGradientBoostingRegressor, X_regression, y_regression)
+])
+def test_warm_start_yields_identical_results(GradientBoosting, X, y):
+    # Make sure that fitting 50 iterations and then 25 with warm start is
+    # equivalent to fitting 75 iterations.
+
+    rng = 42
+    gb_warm_start = GradientBoosting(
+        n_iter_no_change=100, max_iter=50, random_state=rng, warm_start=True
+    )
+    gb_warm_start.fit(X, y).set_params(max_iter=75).fit(X, y)
+
+    gb_no_warm_start = GradientBoosting(
+        n_iter_no_change=100, max_iter=75, random_state=rng, warm_start=False
+    )
+    gb_no_warm_start.fit(X, y)
+
+    # Check identical nodes for each tree
+    for (pred_ith_warm, pred_ith_no_warm) in zip(gb_warm_start._predictors,
+                                                 gb_no_warm_start._predictors):
+        for (predictor_warm, predictor_no_warm) in zip(pred_ith_warm,
+                                                       pred_ith_no_warm):
+            np.testing.assert_array_equal(
+                predictor_warm.nodes,
+                predictor_no_warm.nodes
+            )
+
+    # Check identical predictions
+    np.testing.assert_allclose(gb_warm_start.predict(X),
+                               gb_no_warm_start.predict(X))
+
+
+def test_warm_start_max_depth():
+    # Test if possible to fit trees of different depth in ensemble.
+
+    X_classification, y_classification = make_classification(
+        n_samples=1000, n_features=100, n_informative=70, random_state=0)
+    X_regression, y_regression = make_regression(
+        n_samples=1000, n_features=100, n_informative=70, random_state=0)
+
+    for (GradientBoosting, X, y) in zip(
```

Comment:
Please parametrize instead:


```py

@pytest.mark.parametrize('GradientBoosting, X, y', [
    (HistGradientBoostingClassifier, X_classification, y_classification),
    (HistGradientBoostingRegressor, X_regression, y_regression)
])
```

Reply:
I have to generate other datasets for this one (we need datasets with more relevant features, otherwise `max_depth` is always equal to 2 and I cannot test this).

Reply:
what if you set `min_samples_leaf` to 1?

Reply:
It works! I'm going to use that.